File: extensions/midi/libmidi.m
@import Foundation;
@import Cocoa;
@import LuaSkin;

static const char *USERDATA_TAG = "hs.midi";

static int midi_removed(lua_State *L) {
    return luaL_error(L, "hs.midi has been removed (MIKMIDI dependency dropped)");
}

static const luaL_Reg moduleLib[] = {
    {"new", midi_removed},
    {"newVirtualSource", midi_removed},
    {"devices", midi_removed},
    {"virtualSources", midi_removed},
    {"deviceCallback", midi_removed},
    {NULL, NULL}
};

static const luaL_Reg module_metaLib[] = {
    {NULL, NULL}
};

int luaopen_hs_libmidi(lua_State* L) {
    LuaSkin *skin = [LuaSkin sharedWithState:L];
    [skin registerLibrary:USERDATA_TAG functions:moduleLib metaFunctions:module_metaLib];
    return 1;
}

File: extensions/midi/libmidi.swift
import Foundation
import Cocoa
import LuaSkin

private let USERDATA_TAG = "hs.midi"

private func midi_removed(_ L: OpaquePointer!) -> Int32 {
    return luaL_error(L, "hs.midi has been removed (MIKMIDI dependency dropped)")
}

private var moduleLib: [luaL_Reg] = [
    luaL_Reg(name: strdup("new"),              func: midi_removed),
    luaL_Reg(name: strdup("newVirtualSource"), func: midi_removed),
    luaL_Reg(name: strdup("devices"),          func: midi_removed),
    luaL_Reg(name: strdup("virtualSources"),   func: midi_removed),
    luaL_Reg(name: strdup("deviceCallback"),   func: midi_removed),
    luaL_Reg(name: nil,                        func: nil),
]

private var module_metaLib: [luaL_Reg] = [
    luaL_Reg(name: nil, func: nil),
]

@_cdecl("luaopen_hs_libmidi")
func luaopen_hs_libmidi(_ L: OpaquePointer!) -> Int32 {
    let skin = LuaSkin.shared(withState: L)!
    skin.registerLibrary(USERDATA_TAG, functions: &moduleLib, metaFunctions: &module_metaLib)
    return 1
}
